homeassistant/components/mqtt/client.py

```diff
@@ -543,21 +641,27 @@ def _client_unsubscribe(topic: str) -> int:
             _raise_on_error(result)
             return mid
 
-        async with self._paho_lock:
-            if self._is_active_subscription(topic):
-                # Other subscriptions on topic remaining - don't unsubscribe.
+        if self._is_active_subscription(topic):
+            subs = self._matching_subscriptions(topic)
+            if self._max_qos[topic] == 0:
```

Comment:
Move up one line.  No need to even do the matching of subscriptions (which itself is an interation) (ie. get subs) if max_qos == 0, since subs is only used after if this didn't trigger a return.